drivers/input/keyboard/keyboard_gpio.c

```diff
@@ -0,0 +1,408 @@
+	.longpress_ms = DT_PROP_OR(node_id, longpress_ms, 0),	\
+	.holdpress_ms = DT_PROP_OR(node_id, holdpress_ms, 0),	\
+}
+
+typedef enum {
+	KEY_STATE_NONE,
+	KEY_STATE_PRESSED,
+	KEY_STATE_LONG_PRESSED,
+	KEY_STAT_MAXNBR
+} key_state_t;
+
+typedef enum {
+	KEY_EVENT_RELEASE = KEY_RELEASE,
+	KEY_EVENT_PRESSED = KEY_PRESSED,
+	KEY_EVENT_LONG_PRESSED = KEY_LONG_PRESSED,
+	KEY_EVENT_HOLD_PRESSED = KEY_HOLD_PRESSED,
+	KEY_EVENT_LONG_RELEASE = KEY_LONG_RELEASE,
+	KEY_EVENT_NONE = 0xff,
+} key_event_t;
+
+struct kbd_gpio_driver {
+	struct gpio_callback gpio_data;
+
+	void *pdata;
+	bool pressed;
+	key_state_t state;
+	key_event_t event;
+
+	uint32_t count_debounce;
+	uint32_t count_cycle;
+};
+
+struct kbd_gpio_data {
+	const struct device *dev;
+	struct input_dev *input;
+	struct k_work_delayable delayed_work;
+	struct kbd_gpio_driver *driver;
+};
+
+struct kbd_gpio_config {
+	uint8_t num_keys;
+	const struct gpio_dt_spec *gpio;
+	const struct key_info_dt_spec *info;
+};
+
+#define MS_TO_CYCLE(ms, default)      (((ms) ? (ms) : (default)) / SCAN_INTERVAL)
+#define KEY_EVENT_CODE(code, default) ((code != KEY_RESERVED) ? code : default)
+#define KEY_EVENT_CALL(input, code, event)		\
+{												\
+	input_report_key(input, code, event);		\
+}
+
+static bool kbd_gpio_one_key_proc(const struct device *dev, int idx)
+{
+	const struct kbd_gpio_config *config = dev->config;
+	struct kbd_gpio_data *data = dev->data;
+	struct input_dev *input = data->input;
+
+	const struct gpio_dt_spec *gpio = &config->gpio[idx];
+	const struct key_info_dt_spec *info = &config->info[idx];
+	struct kbd_gpio_driver *driver = &data->driver[idx];
+
+	bool pressed = gpio_pin_get(gpio->port, gpio->pin);
+
+	if (driver->state != KEY_STATE_NONE) {
+		driver->count_cycle++;
+	}
+
+	if (driver->pressed != pressed) {
+		if (driver->count_debounce++ >= MS_TO_CYCLE(info->debounce_ms, TIME_DEBOUNCE)) {
+			driver->pressed = pressed;
+			driver->count_debounce = 0;
+		}
+	} else {
+		driver->count_debounce = 0;
+	}
+
+	switch (driver->state) {
+	case KEY_STATE_NONE:
+		if (driver->pressed == true) {
+			driver->event = KEY_EVENT_PRESSED;
+			driver->state = KEY_STATE_PRESSED;
+			driver->count_cycle = 0;
+			KEY_EVENT_CALL(input, KEY_EVENT_CODE(info->code, (idx + 1)), driver->event);
+		} else {
+			driver->event = KEY_EVENT_NONE;
+		}
+		break;
+
+	case KEY_STATE_PRESSED:
+		if (driver->pressed == false) {
+			driver->event = KEY_EVENT_RELEASE;
+			driver->state = KEY_STATE_NONE;
+			KEY_EVENT_CALL(input, KEY_EVENT_CODE(info->code, (idx + 1)), driver->event);
+		} else if (MS_TO_CYCLE(info->longpress_ms, TIME_LONGPRESS) > 0 &&
+			   driver->count_cycle >= MS_TO_CYCLE(info->longpress_ms, TIME_LONGPRESS)) {
+			driver->event = KEY_EVENT_LONG_PRESSED;
+			driver->state = KEY_STATE_LONG_PRESSED;
+			driver->count_cycle = 0;
+			KEY_EVENT_CALL(input, KEY_EVENT_CODE(info->code, (idx + 1)), driver->event);
+		}
+		break;
+
+	case KEY_STATE_LONG_PRESSED:

```

Comment:
I suggest separating interaction events from button handler - make things ready for future expansion.
Here is code from NCS (again based on event manager) https://github.com/nrfconnect/sdk-nrf/blob/main/subsys/caf/modules/click_detector.c